xdsl/utils/scoped_dict.py

```diff
@@ -0,0 +1,62 @@
+from __future__ import annotations
+
+from typing import Generic, TypeVar
+
+_Key = TypeVar("_Key")
+_Value = TypeVar("_Value")
+
+
+class ScopedDict(Generic[_Key, _Value]):
+    """
+    A tiered mapping from keys to values.
+    Once a value is set for a key, it cannot be overwritten.
+    A ScopedDict may have a parent dict, which is used as a fallback when a value for a
+    key is not found.
+    If a ScopedDict and its parent have values for the same key, the child value will be
+    returned.
+    This structure is useful for contexts where keys and values have a known scope, such
+    as during IR construction from an Abstract Syntax Tree.
+    ScopedDict instances may have a `name` property as a hint during debugging.
+    """
+
+    _local_scope: dict[_Key, _Value]
+    parent: ScopedDict[_Key, _Value] | None
+    name: str | None
+
+    def __init__(
+        self,
+        parent: ScopedDict[_Key, _Value] | None = None,
+        *,
+        name: str | None = None,
+    ) -> None:
+        self._local_scope = {}
+        self.parent = parent
+        self.name = name
+
+    def __getitem__(self, key: _Key) -> _Value:
+        """
+        Fetch key from environment. Attempts to first fetch from current scope,
+        then from parent scopes. Raises KeyError error if not found.
+        """
+        local = self._local_scope.get(key)
+        if local is not None:
+            return local
+        if self.parent is None:
+            raise KeyError()
+        return self.parent[key]
+
+    def __setitem__(self, key: _Key, value: _Value):
+        """
+        Assign key to current scope. Raises InterpretationError if key already
+        assigned to.
+        """
+        if key in self._local_scope:
+            raise ValueError(
```

Comment:
Should this be a `KeyError`?

Reply:
I considered it, but I have a feeling that this is not the intended purpose of KeyError, which to me signals that something does not exist for a given key, whereas this error is raised when something does already exist. In either case, I'm not sure what the user can do to recover.